File: avl-tree-list/avltreelist.py
# 
# AVL tree list (Python)
# 
# Copyright (c) 2021 Project Nayuki. (MIT License)
# https://www.nayuki.io/page/avl-tree-list
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
# - The above copyright notice and this permission notice shall be included in
#   all copies or substantial portions of the Software.
# - The Software is provided "as is", without warranty of any kind, express or
#   implied, including but not limited to the warranties of merchantability,
#   fitness for a particular purpose and noninfringement. In no event shall the
#   authors or copyright holders be liable for any claim, damages or other
#   liability, whether in an action of contract, tort or otherwise, arising from,
#   out of or in connection with the Software or the use or other dealings in the
#   Software.
# 

from __future__ import annotations
from typing import cast, ClassVar, Generator, Generic, Iterable, List, Optional, Set, TypeVar


E = TypeVar("E")
T = TypeVar("T")

class AvlTreeList(Generic[E]):
	
	root: AvlTreeList.Node[E]
	size: int
	
	
	def __init__(self, lst: Optional[Iterable[E]] = None):
		self.clear()
		if lst is not None:
			self.extend(lst)
	
	
	def __len__(self) -> int:
		return self.root.size
	
	
	def __getitem__(self, index: int) -> E:
		if not (0 <= index < len(self)):
			raise IndexError()
		return self.root.get_node_at(index).value
	
	
	def __setitem__(self, index: int, val: E) -> None:
		if not (0 <= index < len(self)):
			raise IndexError()
		self.root.get_node_at(index).value = val
	
	
	def insert(self, index: int, val: E) -> None:
		if not (0 <= index <= len(self)):  # Different constraint than the other methods
			raise IndexError()
		self.root = self.root.insert_at(index, val)
	
	
	def append(self, val: E) -> None:
		self.insert(len(self), val)
	
	
	def extend(self, lst: Iterable[E]) -> None:
		for val in lst:
			self.append(val)
	
	
	def pop(self, index: Optional[int] = None) -> E:
		if index is None:
			index = len(self) - 1
		result: E = self[index]
		del self[index]
		return result
	
	
	def __delitem__(self, index: int) -> None:
		if not (0 <= index < len(self)):
			raise IndexError()
		self.root = self.root.remove_at(index)
	
	
	def clear(self) -> None:
		self.root = cast(AvlTreeList.Node[E], AvlTreeList.Node.EMPTY_LEAF)
	
	
	# Note: Not fail-fast on concurrent modification.
	def __iter__(self) -> Generator[E,None,None]:
		stack: List[AvlTreeList.Node[E]] = []
		node: AvlTreeList.Node[E] = self.root
		while True:
			while node is not cast(AvlTreeList.Node[E], AvlTreeList.Node.EMPTY_LEAF):
				stack.append(node)
				node = _non_none(node.left)
			if len(stack) == 0:
				break
			node = stack.pop()
			yield node.value
			node = _non_none(node.right)
	
	
	def __str__(self) -> str:
		return "[" + ", ".join(str(x) for x in self) + "]"
	
	
	# For unit tests
	def check_structure(self) -> None:
		self.root.check_structure(set())
	
	
	
	class Node(Generic[T]):
		
		EMPTY_LEAF: ClassVar[AvlTreeList.Node[object]]
		
		
		value: T
		height: int
		size: int
		left: Optional[AvlTreeList.Node[T]]
		right: Optional[AvlTreeList.Node[T]]
		
		
		def __init__(self, val: T, isleaf: bool = False):
			# The object stored at this node. Can be None.
			self.value = val
			
			if isleaf:  # For the singleton empty leaf node
				self.height = 0
				self.size   = 0
				self.left   = None
				self.right  = None
				
			else:  # Normal non-leaf nodes
				# The height of the tree rooted at this node. Empty nodes have height 0.
				# This node has height equal to max(left.height, right.height) + 1.
				self.height = 1
				
				# The number of non-empty nodes in the tree rooted at this node, including this node.
				# Empty nodes have size 0. This node has size equal to left.size + right.size + 1.
				self.size   = 1
				
				# The root node of the left subtree.
				self.left   = self._empty_leaf()
				
				# The root node of the right subtree.
				self.right  = self._empty_leaf()
		
		
		def get_node_at(self, index: int) -> AvlTreeList.Node[T]:
			assert 0 <= index < self.size  # Automatically implies self != EMPTY_LEAF, because EMPTY_LEAF.size == 0
			leftsize: int = _non_none(self.left).size
			if index < leftsize:
				return _non_none(self.left).get_node_at(index)
			elif index > leftsize:
				return _non_none(self.right).get_node_at(index - leftsize - 1)
			else:
				return self
		
		
		def insert_at(self, index: int, obj: T) -> AvlTreeList.Node[T]:
			assert 0 <= index <= self.size
			if self is self._empty_leaf():  # Automatically implies index == 0, because EMPTY_LEAF.size == 0
				return AvlTreeList.Node(obj)
			leftsize: int = _non_none(self.left).size
			if index <= leftsize:
				self.left = _non_none(self.left).insert_at(index, obj)
			else:
				self.right = _non_none(self.right).insert_at(index - leftsize - 1, obj)
			self._recalculate()
			return self._balance()
		
		
		def remove_at(self, index: int) -> AvlTreeList.Node[T]:
			assert 0 <= index < self.size  # Automatically implies self != EMPTY_LEAF, because EMPTY_LEAF.size == 0
			EMPTY: AvlTreeList.Node[T] = self._empty_leaf()
			leftsize: int = _non_none(self.left).size
			if index < leftsize:
				self.left = _non_none(self.left).remove_at(index)
			elif index > leftsize:
				self.right = _non_none(self.right).remove_at(index - leftsize - 1)
			elif self.left is EMPTY and self.right is EMPTY:
				return EMPTY
			elif self.left is not EMPTY and self.right is EMPTY:
				return _non_none(self.left)
			elif self.left is EMPTY and self.right is not EMPTY:
				return _non_none(self.right)
			else:
				# Find successor node. (Using the predecessor is valid too.)
				temp: AvlTreeList.Node[T] = _non_none(self.right)
				while temp.left is not self._empty_leaf():
					temp = _non_none(temp.left)
				self.value = temp.value  # Replace value by successor
				self.right = _non_none(self.right).remove_at(0)  # Remove successor node
			self._recalculate()
			return self._balance()
		
		
		def __str__(self) -> str:
			return f"AvlTreeNode(size={self.size}, height={self.height}, val={self.value})"
		
		
		# Balances the subtree rooted at this node and returns the new root.
		def _balance(self) -> AvlTreeList.Node[T]:
			bal: int = self._get_balance()
			assert abs(bal) <= 2
			result: AvlTreeList.Node[T] = self
			if bal == -2:
				assert abs(_non_none(self.left)._get_balance()) <= 1
				if _non_none(self.left)._get_balance() == +1:
					self.left = _non_none(self.left)._rotate_left()
				result = self._rotate_right()
			elif bal == +2:
				assert abs(_non_none(self.right)._get_balance()) <= 1
				if _non_none(self.right)._get_balance() == -1:
					self.right = _non_none(self.right)._rotate_right()
				result = self._rotate_left()
			assert abs(result._get_balance()) <= 1
			return result
		
		
		# 
		#   A            B
		#  / \          / \
		# 0   B   ->   A   2
		#    / \      / \
		#   1   2    0   1
		# 
		def _rotate_left(self) -> AvlTreeList.Node[T]:
			if self.right is self._empty_leaf():
				raise ValueError()
			root: AvlTreeList.Node[T] = _non_none(self.right)
			self.right = root.left
			root.left = self
			self._recalculate()
			root._recalculate()
			return root
		
		
		# 
		#     B          A
		#    / \        / \
		#   A   2  ->  0   B
		#  / \            / \
		# 0   1          1   2
		# 
		def _rotate_right(self) -> AvlTreeList.Node[T]:
			if self.left is self._empty_leaf():
				raise ValueError()
			root: AvlTreeList.Node[T] = _non_none(self.left)
			self.left = root.right
			root.right = self
			self._recalculate()
			root._recalculate()
			return root
		
		
		# Needs to be called every time the left or right subtree is changed.
		# Assumes the left and right subtrees have the correct values computed already.
		def _recalculate(self) -> None:
			left: AvlTreeList.Node[T] = _non_none(self.left)
			right: AvlTreeList.Node[T] = _non_none(self.right)
			assert self is not self._empty_leaf()
			assert left.height >= 0 and right.height >= 0
			assert left.size >= 0 and right.size >= 0
			self.height = max(left.height, right.height) + 1
			self.size = left.size + right.size + 1
			assert self.height >= 0 and self.size >= 0
		
		
		def _get_balance(self) -> int:
			return _non_none(self.right).height - _non_none(self.left).height
		
		
		# For unit tests, invokable by the outer class.
		def check_structure(self, visitednodes: Set[AvlTreeList.Node[T]]) -> None:
			if self is self._empty_leaf():
				return
			
			if self in visitednodes:
				raise AssertionError("AVL tree structure violated: Not a tree")
			visitednodes.add(self)
			_non_none(self.left ).check_structure(visitednodes)
			_non_none(self.right).check_structure(visitednodes)
			
			if self.height != max(_non_none(self.left).height, _non_none(self.right).height) + 1:
				raise AssertionError("AVL tree structure violated: Incorrect cached height")
			if self.size != _non_none(self.left).size + _non_none(self.right).size + 1:
				raise AssertionError("AVL tree structure violated: Incorrect cached size")
			if abs(self._get_balance()) > 1:
				raise AssertionError("AVL tree structure violated: Height imbalance")
		
		
		def _empty_leaf(self) -> AvlTreeList.Node[T]:
			return cast(AvlTreeList.Node[T], AvlTreeList.Node.EMPTY_LEAF)


# Static initializer. A bit of a hack, but more elegant than using None values as leaf nodes.
AvlTreeList.Node.EMPTY_LEAF = AvlTreeList.Node(None, True)


def _non_none(val: Optional[AvlTreeList.Node[E]]) -> AvlTreeList.Node[E]:
	if val is None:
		raise ValueError()
	return val

File: avl-tree-list/AvlTreeList.ts
/* 
 * AVL tree list (TypeScript)
 * 
 * Copyright (c) 2018 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/avl-tree-list
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

"use strict";


/*---- Data structure main class (public) ----*/

class AvlTreeList<E> {
	
	private root: AvlTreeListNode<E> = new AvlTreeListEmptyNode<E>();
	
	
	// Constructs a new AVL tree list. If called with no arguments, this creates a blank list.
	// Otherwise with one argument, this creates a list representing the same elements as the given array.
	// For example:
	//   let a = new AvlTreeList<boolean>();  // Blank, zero-length list
	//   let b = new AvlTreeList<number>([2,7,1,8]);  // Has the four elements 2,7,1,8
	public constructor(arr?: Array<E>) {
		if (arguments.length == 1 && arr !== undefined)
			arr.forEach((val: E) => this.push(val));
		else if (arguments.length != 0)
			throw "Illegal argument";
	}
	
	
	// The property 'length' returns the length of this list. It is a property with a
	// getter function, so that parentheses are not used to get the value. For example:
	//   let list = new AvlTreeList<number>([3,1,4]);
	//   let n = list.length;  // 3
	public get length(): number {
		return this.root.size;
	}
	
	
	// Returns the element at the given index in this list, where 0 <= index < length.
	// For example: [a,b,c] -> get(2) returns c.
	public get(index: number): E {
		if (index < 0 || index >= this.length)
			throw "Index out of bounds";
		return (this.root as AvlTreeListInternalNode<E>).getNodeAt(index).value;
	}
	
	
	// Sets the element at the given index in this list to the given value, where 0 <= index < length.
	// For example: [a,b,c] -> set(1, f) -> [a,f,c].
	public set(index: number, val: E): void {
		if (index < 0 || index >= this.length)
			throw "Index out of bounds";
		(this.root as AvlTreeListInternalNode<E>).getNodeAt(index).value = val;
	}
	
	
	// Appends the given element to the end of this list.
	// For example: [a,b] -> push(c) -> [a,b,c].
	public push(val: E): void {
		this.root = this.root.insertAt(this.length, val);
	}
	
	
	// Inserts the given element to the given index in this list, where 0 <= index <= length.
	// For example: [a,b,c] -> insert(0, e) -> [e,a,b,c] -> insert(4, d) -> [e,a,b,c,d].
	public insert(index: number, val: E): void {
		if (index < 0 || index > this.length)  // Different constraint than the other methods
			throw "Index out of bounds";
		this.root = this.root.insertAt(index, val);
	}
	
	
	// Removes the element at the given index in this list, shifting all later elements forward by one index.
	// For example: [a,b,c,d] -> remove(1) -> [a,c,d].
	public remove(index: number): void {
		if (index < 0 || index >= this.length)
			throw "Index out of bounds";
		this.root = (this.root as AvlTreeListInternalNode<E>).removeAt(index);
	}
	
	
	// Removes and returns the first element of this list.
	// For example: [a,b,c] -> shift() returns a -> [b,c].
	public shift(): E {
		if (this.length == 0)
			throw "List is empty";
		const result: E = this.get(0);
		this.remove(0);
		return result;
	}
	
	
	// Removes and returns the last element of this list.
	// For example: [a,b,c] -> pop() returns c -> [a,b].
	public pop(): E {
		const len: number = this.length;
		if (len == 0)
			throw "List is empty";
		const result = this.get(len - 1);
		this.remove(len - 1);
		return result;
	}
	
	
	// Removes all elements from this list, resetting the length to zero.
	// For example: [a,b,c] -> clear() -> [].
	public clear(): void {
		this.root = new AvlTreeListEmptyNode<E>();
	}
	
	
	// Returns a shallow copy of the given subrange of this list. Both arguments are optional - a missing end
	// means the end of the list, and a missing start means the beginning of the list. The returned list has its
	// own unique data structure and node objects, but the underlying values being stored are still the same objects.
	// Note for convenience that lst.slice() returns a clone of the entire list.
	public slice(start?: number, end?: number) {
		if (arguments.length < 2)
			end = this.length;
		end = end as number;
		if (arguments.length < 1)
			start = 0;
		start = start as number;
		if (start < 0)
			start += this.length;
		if (end < 0)
			end += this.length;
		let result = new AvlTreeList<E>();
		for (let i = start; i < end; i++)
			result.push(this.get(i));
		return result;
	}
	
	
	// Removes 'count' elements starting at index 'start', then inserts the rest of the arguments
	// at index 'start', and finally returns an AvlTreeList of the removed elements.
	public splice(start: number, count: number, ...vals: Array<E>): AvlTreeList<E> {
		// Clamp start index and count
		if (start > this.length)
			start = this.length;
		else if (start < 0) {
			start += this.length;
			if (start < 0)
				start = 0;
		}
		count = Math.min(this.length - start, count);
		
		let result = new AvlTreeList<E>();
		for (let i = 0; i < count; i++) {
			const val: E = this.get(start);
			result.push(val);
			this.remove(start);
		}
		vals.forEach((val: E, i: number) => this.insert(start + i, val));
		return result;
	}
	
	
	// Calls func(val, index) on each element of this list in sequential order.
	// The optional 'thisArg' sets the value of 'this' when calling the given function.
	public forEach(func: (x: E, i: number) => any, thisArg?: any): void {
		for (let iter = this.iterator(), i = 0; iter.hasNext(); i++)
			func.call(thisArg, iter.next(), i);
	}
	
	
	// Returns an iterator object to traverse the entire list from beginning to end. The iterator has two methods
	// (modelled after java.util.Iterator): hasNext() returning a boolean, and next() returning an element value.
	// When processing the whole list, using an iterator is more efficient since it uses O(n) total time
	// (so O(1) amortized but O(log n) worst-case time per element), whereas looping over all indexes
	// and calling get() on each one takes O(n log n) total time. For example:
	//   let list = (...);  // AvlTreeList object
	//   let iter = list.iterator();
	//   while (iter.hasNext()) {
	//     let val = iter.next();
	//     (... do something with val ...)
	//   }
	public iterator(): AvlTreeListIterator<E> {
		return new AvlTreeListIterator<E>(this.root);
	}
	
	
	// Returns the contents of this list as a JavaScript array.
	// For example:
	//   let list = new AvlTreeList<string>();
	//   list.push("x");
	//   list.insert(0, "a");
	//   list.push("c");
	//   list.set(1, "b");
	//   let a = list.toArray();  // ["a","b","c"]
	public toArray(): Array<E> {
		let result: Array<E> = [];
		for (let iter = this.iterator(); iter.hasNext(); )
			result.push(iter.next());
		return result;
	}
	
	
	// For unit tests. Returns nothing or throws an exception.
	public checkStructure(): void {
		this.root.checkStructure();
	}
	
}



/*---- Node helper classes (private) ----*/

interface AvlTreeListNode<E> {
	
	// The height of the tree rooted at this node.
	height: number;
	
	// The number of non-empty nodes in the tree rooted at this node, including this node.
	size: number;
	
	
	insertAt(index: number, obj: E): AvlTreeListInternalNode<E>;
	
	checkStructure(): void;
	
}


class AvlTreeListEmptyNode<E> implements AvlTreeListNode<E> {
	
	public readonly height = 0;
	
	public readonly size = 0;
	
	
	public insertAt<E>(index: number, obj: E): AvlTreeListInternalNode<E> {
		if (index == 0)
			return new AvlTreeListInternalNode<E>(obj, this);
		else
			throw "Index out of bounds";
	}
	
	
	public checkStructure(): void {}
	
}


class AvlTreeListInternalNode<E> implements AvlTreeListNode<E> {
	
	// This node has height equal to max(left.height, right.height) + 1.
	public height: number = 1;
	
	// This node has size equal to left.size + right.size + 1.
	// Public for the sake of AvlTreeList.
	public size: number = 1;
	
	// The root node of the left  subtree. Public for the sake of AvlTreeListIterator.
	public left : AvlTreeListNode<E>;
	
	// The root node of the right subtree. Public for the sake of AvlTreeListIterator.
	public right: AvlTreeListNode<E>;
	
	
	public constructor(
			// The object stored at this node. Can be any JavaScript type.
			// Public for the sake of AvlTreeList.
			public value: E,
			empty: AvlTreeListEmptyNode<E>) {
		this.left  = empty;
		this.right = empty;
	}
	
	
	private get leftNode(): AvlTreeListInternalNode<E> {
		if (this.left.size == 0)
			throw "Assertion error";
		return this.left as AvlTreeListInternalNode<E>;
	}
	
	
	private get rightNode(): AvlTreeListInternalNode<E> {
		if (this.right.size == 0)
			throw "Assertion error";
		return this.right as AvlTreeListInternalNode<E>;
	}
	
	
	public getNodeAt(index: number): AvlTreeListInternalNode<E> {
		if (index < 0 || index >= this.size)
			throw "Assertion error";
		const leftSize: number = this.left.size;
		if (index < leftSize)
			return this.leftNode.getNodeAt(index);
		else if (index > leftSize)
			return this.rightNode.getNodeAt(index - leftSize - 1);
		else
			return this;
	}
	
	
	public insertAt(index: number, obj: E): AvlTreeListInternalNode<E> {
		if (index < 0 || index > this.size)
			throw "Assertion error";
		const leftSize: number = this.left.size;
		if (index <= leftSize)
			this.left = this.left.insertAt(index, obj);
		else
			this.right = this.right.insertAt(index - leftSize - 1, obj);
		this.recalculate();
		return this.balance();
	}
	
	
	public removeAt(index: number): AvlTreeListNode<E> {
		if (index < 0 || index >= this.size)
			throw "Assertion error";
		const leftSize: number = this.left.size;
		if (index < leftSize)
			this.left = this.leftNode.removeAt(index);
		else if (index > leftSize)
			this.right = this.rightNode.removeAt(index - leftSize - 1);
		else if (this.left.size == 0 && this.right.size == 0)
			return this.left;  // Empty
		else if (this.left.size != 0 && this.right.size == 0)
			return this.left;
		else if (this.left.size == 0 && this.right.size != 0)
			return this.right;
		else {
			// Find successor node. (Using the predecessor is valid too.)
			let temp = this.rightNode;
			while (temp.left.size != 0)
				temp = temp.leftNode;
			this.value = temp.value;  // Replace value by successor
			this.right = this.rightNode.removeAt(0);  // Remove successor node
		}
		this.recalculate();
		return this.balance();
	}
	
	
	// Balances the subtree rooted at this node and returns the new root.
	private balance(): AvlTreeListInternalNode<E> {
		const bal: number = this.getBalance();
		if (Math.abs(bal) > 2)
			throw "Assertion error";
		let result: AvlTreeListInternalNode<E> = this;
		if (bal == -2) {
			let left = this.leftNode;
			if (Math.abs(left.getBalance()) > 1)
				throw "Assertion error";
			if (left.getBalance() == +1)
				this.left = left.rotateLeft();
			result = this.rotateRight();
		} else if (bal == +2) {
			let right = this.rightNode;
			if (Math.abs(right.getBalance()) > 1)
				throw "Assertion error";
			if (right.getBalance() == -1)
				this.right = right.rotateRight();
			result = this.rotateLeft();
		}
		if (Math.abs(result.getBalance()) > 1)
			throw "Assertion error";
		return result;
	}
	
	
	/* 
	 *   A            B
	 *  / \          / \
	 * 0   B   ->   A   2
	 *    / \      / \
	 *   1   2    0   1
	 */
	private rotateLeft(): AvlTreeListInternalNode<E> {
		if (this.right.size == 0)
			throw "Assertion error";
		let root = this.rightNode;
		this.right = root.left;
		root.left = this;
		this.recalculate();
		root.recalculate();
		return root;
	}
	
	
	/* 
	 *     B          A
	 *    / \        / \
	 *   A   2  ->  0   B
	 *  / \            / \
	 * 0   1          1   2
	 */
	private rotateRight(): AvlTreeListInternalNode<E> {
		if (this.left.size == 0)
			throw "Assertion error";
		let root = this.leftNode;
		this.left = root.right;
		root.right = this;
		this.recalculate();
		root.recalculate();
		return root;
	}
	
	
	// Needs to be called every time the left or right subtree is changed.
	// Assumes the left and right subtrees have the correct values computed already.
	private recalculate(): void {
		if (this.size == 0)
			throw "Assertion error";
		if (this.left.height < 0 || this.right.height < 0)
			throw "Assertion error";
		if (this.left.size < 0 || this.right.size < 0)
			throw "Assertion error";
		this.height = Math.max(this.left.height, this.right.height) + 1;
		this.size = this.left.size + this.right.size + 1;
		if (this.height < 0 || this.size < 0)
			throw "Assertion error";
	}
	
	
	private getBalance(): number {
		return this.right.height - this.left.height;
	}
	
	
	// For unit tests, invokable by the main class.
	public checkStructure(): void {
		this.left .checkStructure();
		this.right.checkStructure();
		if (this.height != Math.max(this.left.height, this.right.height) + 1)
			throw "AVL tree structure violated: Incorrect cached height";
		if (this.size != this.left.size + this.right.size + 1)
			throw "AVL tree structure violated: Incorrect cached size";
		if (Math.abs(this.getBalance()) > 1)
			throw "AVL tree structure violated: Height imbalance";
	}
	
}



/*---- Iterator helper class (private) ----*/

// Note: An iterator is not fail-fast on concurrent modification.
class AvlTreeListIterator<E> {
	
	private stack: Array<AvlTreeListInternalNode<E>>;
	
	
	public constructor(root: AvlTreeListNode<E>) {
		this.stack = [];
		let maybeNode: AvlTreeListNode<E> = root;
		while (maybeNode.size != 0) {
			const node = maybeNode as AvlTreeListInternalNode<E>;
			this.stack.push(node);
			maybeNode = node.left;
		}
	}
	
	
	public hasNext(): boolean {
		return this.stack.length > 0;
	}
	
	
	public next(): E {
		if (!this.hasNext())
			throw "No next element";
		
		let node = this.stack.pop() as AvlTreeListInternalNode<E>;  // Never undefined
		const result: E = node.value;
		let maybeNode: AvlTreeListNode<E> = node.right;
		while (maybeNode.size != 0) {
			const node = maybeNode as AvlTreeListInternalNode<E>;
			this.stack.push(node);
			maybeNode = node.left;
		}
		return result;
	}
	
}
